### DISCUSSION TITLE
Service Decoupling

### DISCUSSION THREAD

**Discussion by ernst86913** [directus/directus#16995] [Feature Requests]
Hi - currently you have several occurrences where one service calls another one - eg login Service directly calls activity service and similar.

Wouldn’t it be better from an architecture perspective to decouple those services and instead emit an event where the other service then reacts on it and performs the required actions?

in the example above: login service throws an event “NewLogin” and the activity service takes this event and registers the login in the activity log.

Or are there some good reasons to have it coupled?

thanks in advance for any insights.

**Comment by paescuj** [comment]
> Or are there some good reasons to have it coupled?

Let me ask a counter question 😃 What advantages do you see exactly in decoupling? (if we leave out the architecture perspective)

**Reply by ernst86913** [reply]
Example: inviteUser… this directly includes the handling of mail messages… imagine you don’t want to send via mail… but via different mechanisms… as of a sudden you end up programming a whole “message center” concept with options into the user service. Now in parallel you have notifications… also there you send a mail… now imaging you want to send push notifications instead… or SMS… or… now you end up doing the next „message center“ in notifications service as well.

Better way would be to focus each service on what it is for… without interdependencies to others… and just throw an event „user.newInvite“ or „notification.send“… and a central service „massage center service“ is listening for the event… and based on project/user settings sends it to the right channel.

Now think of wanting a custom module which does something completely different when a notification is sent… it could listen to the event and could hook into it… currently that’s not possible…

… just some stupid examples - I hope it illustrates the advantages

**Reply by rijkvanzanten** [reply]
Doesn't that make the flow of data way more opaque though? Right now, you can follow the flow of data and operations executed against that data from "the top down" when looking at any particular method; when you're relying on an event bus where other services register listeners, it becomes way harder to figure out what is happening where and when (and it's also a tad trickier to control the _order_ of these callbacks, in case services have to rely on each other's results). That being said, I _do_ agree that certain bits in the API codebase can be decoupled and made into individual functions instead of services (f.e. the authorization service could just be a set of functions rather than a single service class) (cc @jaads)

**Reply by ernst86913** [reply]
Yes - I agree - it does… but the benefits of better maintainability and especially better extensibility usually outweigh those challenges. You will see when the code base grows… and - for more complex Workflows you anyway already have a workflow tool built-in, which makes activities traceable.

BTW - in some cases the direct calls still make sense - eg writing revisions or to the activity log… but notifications and messages can be very complex with lots of different channels… I am sure at some point a separate „message center“ module will be needed…

**Reply by paescuj** [reply]
Hmm, I also agree that it could make sense to decouple certain bits of the code base. For example, if the sole purpose is to inform other entities (for example, a service which handles notifications, as you mentioned) where we are not interested in its result at all, we could indeed just emit an event.

For other parts (core services, ...) however, I hold the opinion that having it coupled is the picture-perfect (or "pattern-perfect" 😉) way to go. Especially for the reason that the services _depend_ on each other, as @rijkvanzanten has already argued:
> [...] and it's also a tad trickier to control the order of these callbacks, in case services have to rely on each other's results [...]

**Reply by ernst86913** [reply]
Makes sense for me... just wanted to raise this as a point to avoid future pains... for some topics it definitely would make sense... and for some not... things are never black and white... ;-)

**Reply by paescuj** [reply]
Yeah, I think it's precious to have a discussion on such things! Thank you for raising this! 👍 

> things are never black and white... ;-)

Very true 😃

**Comment by jaads** [comment]
Nice discussion :) 
> things are never black and white

Soo true, also regarding coupling actually since it's more of a degree than a simple yes or no. I think the services are already decoupled to a suitable extend. It would not be worth it to refactor the engine to a event driven architecture just to have the services as loosely coupled as possible. It would also get a little complicated due to the fact the services rely on each other as @rijkvanzanten mentioned. 

But regardless of the way modules interact with each other (indirect or direct) it is of course always good to separate logic into modules or functions for better maintainability. In my opinion an event driven approach doesn't target maintainability and modularity directly but rather targets flexibility. If we e.g. would add more features like push or sms notifications we should indeed build some sort of message center instead of having all the logic implemented across various services. However, this can also be done with direct calls and can still result in a good maintainable design.

**Reply by ernst86913** [reply]
Basically fully agree... the only thing to keep in mind regarding 

"If we e.g. would add more features like push or sms notifications we should indeed build some sort of message center instead of having all the logic implemented across various services. However, this can also be done with direct calls and can still result in a good maintainable design."

... keep in mind that you want an extensible architecture where users want to add specific extensions based on their use cases...

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
